### DISCUSSION TITLE
Generics for dynamically adding filters to boxed queries containing joins

### DISCUSSION THREAD

**Discussion by sunny-g** [diesel-rs/diesel#3499] [Q&A]
Somewhat piggybacking off of https://github.com/diesel-rs/diesel/discussions/3468, I have the following simplified schema:

```rust
diesel::table! {
    stac (stac_id) {
        stac_id -> Uuid,
        observed -> Timestamptz,
    }
}

diesel::table! {
    eo (eo_id) {
        eo_id -> Uuid,
        stac_id -> Nullable<Uuid>,
        gsd -> Nullable<Float8>,
    }
}

diesel::joinable!(eo -> stac (stac_id));
```

and have request and filter types designed to map directly to queries against this schema, i.e.:

```rust
pub struct StacRequest {
    id: Option<StringFilter>,
    eo: Option<EoRequest>,
}

pub struct EoRequest {
    gsd: Option<DoubleFilter>,
}
```

I've already implemented `Expression`, `AsExpression`, and other traits for the `*Filter` types so that when provided with a column (via a custom `AsFilterExpression<Lhs>` trait), they can produce the appropriate filter expression (Eq, Like, etc).

Putting them together with reference to other discussions about boxed queries, I can get (something similar to) the following code to work fine:

```rust
impl StacRequest {
  fn add_to_query<'a>(self, mut query: dsl::IntoBoxed<'a, stac::table, Pg>) -> dsl::IntoBoxed<'a, stac::table, Pg> {
    if let Some(id_filter) = self.id {
        query = query.filter(id_filter.satisfies(stac::id))
    }

    query
  }
}
```

The problem occurs when I try putting this function into a method/trait on the `*Request` type *and* making it generic over the query source (such that I can add `EoRequest` filters to a query against `eo::table`, and use the same method to add filters to a query against `stac::table.inner_join(eo::table)`.

I've tried half-dozen different ways of organizing the filters, joins, and box'ings, and the problem ultimately boils down to not knowing how to make a function/method akin to this with the appropriate generics:

```rust
impl StacRequest {
  fn add_to_query<X>(self, mut query: X) -> X {
    if let Some(id_filter) = self.id {
        query = query.filter(id_filter.satisfies(stac::id))
    }

    // type would likely require a join regardless of the request specifying it or not
    query = query.inner_join(eo::table);
    if let Some(eo_filter) = self.eo {
        query = eo_filter.add_to_query(query);
    }

    query
  }
}
```

Granted, in this specific case, `X` could be `dsl::IntoBoxed<'a, InnerJoin<stac::table, eo::table>, Pg>`; however, I've omitted the other dozen tables that can be joined to `stac` and joined via a `StacRequest` from this simplified example and so (without resorting to a ton of macros), I'd like to just rely on trait bounds if possible.

Any guidance or pointers to examples or documentation I can reference?

**Comment by weiznich** [comment]
As already written in https://github.com/diesel-rs/diesel/discussions/3468: I do not recommend this route, therefore I do also not provide detailed instructions in how to implement this.

Generally speaking: You must satisfy any bounds requested by the compiler. Normally you want to look at the `required because …` part of the error messages and use that as bound. You should expect that there are quite a lot of bounds required to make this work.

**Comment by sunny-g** [comment]
Thanks for replying!

> You must satisfy any bounds requested by the compiler. Normally you want to look at the required because … part of the error messages and use that as bound.

Of course, and I tried going down the route of satisfying the bounds, but they explode pretty quickly and because a number of the traits are provided through blanket implementations or are undocumented publicly, its really difficult to follow a type thru the docs to see how they all layer together, or even find parallel types whose structure or behaviour I could emulate.

Are there any reference docs that explain how all the traits fit together and where their bounds are applied in different parts of a program (schema definition, query definition, query execution, etc)?

**Reply by weiznich** [reply]
> Of course, and I tried going down the route of satisfying the bounds, but they explode pretty quickly and because a number of the traits are provided through blanket implementations or are undocumented publicly, its really difficult to follow a type thru the docs to see how they all layer together, or even find parallel types whose structure or behaviour I could emulate.

That's the reason why I do not recommend that approach.

> Are there any reference docs that explain how all the traits fit together and where their bounds are applied in different parts of a program (schema definition, query definition, query execution, etc)?

There is no documentation for these internal traits as they are an internal implementation detail that can change at any point in time. The public API is documented in the API documentation. Everything that's listed there is explained there. Everything that's not listed there is something you should not rely on.

**Reply by sunny-g** [reply]
Ok, thank you for taking the time :)
